// @mysten/kiosk@0.14.4/dist/cjs/types/kiosk.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var kiosk_exports = {};
__export(kiosk_exports, {
  KIOSK_ITEM: () => KIOSK_ITEM,
  KIOSK_LISTING: () => KIOSK_LISTING,
  KIOSK_LOCK: () => KIOSK_LOCK,
  KIOSK_MODULE: () => KIOSK_MODULE,
  KIOSK_OWNER_CAP: () => KIOSK_OWNER_CAP,
  KIOSK_PURCHASE_CAP: () => KIOSK_PURCHASE_CAP,
  KIOSK_TYPE: () => KIOSK_TYPE
});
module.exports = __toCommonJS(kiosk_exports);
const KIOSK_MODULE = "0x2::kiosk";
const KIOSK_TYPE = `${KIOSK_MODULE}::Kiosk`;
const KIOSK_OWNER_CAP = `${KIOSK_MODULE}::KioskOwnerCap`;
const KIOSK_ITEM = `${KIOSK_MODULE}::Item`;
const KIOSK_LISTING = `${KIOSK_MODULE}::Listing`;
const KIOSK_LOCK = `${KIOSK_MODULE}::Lock`;
const KIOSK_PURCHASE_CAP = `${KIOSK_MODULE}::PurchaseCap`;
