// @mysten/kiosk@0.14.4/dist/cjs/types/transfer-policy.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var transfer_policy_exports = {};
__export(transfer_policy_exports, {
  KIOSK_LOCK_RULE: () => KIOSK_LOCK_RULE,
  ROYALTY_RULE: () => ROYALTY_RULE,
  TRANSFER_POLICY_CAP_TYPE: () => TRANSFER_POLICY_CAP_TYPE,
  TRANSFER_POLICY_CREATED_EVENT: () => TRANSFER_POLICY_CREATED_EVENT,
  TRANSFER_POLICY_MODULE: () => TRANSFER_POLICY_MODULE,
  TRANSFER_POLICY_TYPE: () => TRANSFER_POLICY_TYPE
});
module.exports = __toCommonJS(transfer_policy_exports);
const TRANSFER_POLICY_MODULE = "0x2::transfer_policy";
const TRANSFER_POLICY_CREATED_EVENT = `${TRANSFER_POLICY_MODULE}::TransferPolicyCreated`;
const TRANSFER_POLICY_TYPE = `${TRANSFER_POLICY_MODULE}::TransferPolicy`;
const TRANSFER_POLICY_CAP_TYPE = `${TRANSFER_POLICY_MODULE}::TransferPolicyCap`;
const KIOSK_LOCK_RULE = "kiosk_lock_rule::Rule";
const ROYALTY_RULE = "royalty_rule::Rule";
