// @mysten/payment-kit@0.0.17/dist/esm/calls.js
var __typeError = (msg) => {
  throw TypeError(msg);
};
var __accessCheck = (obj, member, msg) => member.has(obj) || __typeError("Cannot " + msg);
var __privateGet = (obj, member, getter) => (__accessCheck(obj, member, "read from private field"), getter ? getter.call(obj) : member.get(obj));
var __privateAdd = (obj, member, value) => member.has(obj) ? __typeError("Cannot add the same private member more than once") : member instanceof WeakSet ? member.add(obj) : member.set(obj, value);
var __privateSet = (obj, member, value, setter) => (__accessCheck(obj, member, "write to private field"), setter ? setter.call(obj, value) : member.set(obj, value), value);
var _packageConfig;
import { coinWithBalance } from "@mysten/sui/transactions";
import {
  createPaymentKey,
  createRegistry,
  deletePaymentRecord,
  processEphemeralPayment,
  processRegistryPayment,
  setConfigEpochExpirationDuration,
  setConfigRegistryManagedFunds,
  withdrawFromRegistry
} from "./contracts/payment_kit/payment_kit.js";
import { getRegistryIdFromName } from "./utils.js";
class PaymentKitCalls {
  constructor(options) {
    __privateAdd(this, _packageConfig);
    /**
     * Creates a `processRegistryPayment` transaction
     *
     * @example
     * ```ts
     * tx.add(lient.paymentKit.call.processRegistryPayment({ nonce, coinType, sender, amount, receiver, registryName }));
     * ```
     */
    this.processRegistryPayment = (options) => {
      const { nonce, coinType, amount, receiver, registryName, registryId } = options;
      const registryIdToUse = registryId ?? getRegistryIdFromName(registryName, __privateGet(this, _packageConfig).namespaceId);
      return processRegistryPayment({
        arguments: {
          registry: registryIdToUse,
          nonce,
          paymentAmount: amount,
          coin: coinWithBalance({
            type: coinType,
            balance: amount
          }),
          receiver
        },
        typeArguments: [coinType]
      });
    };
    /**
     * Creates a `processRegistryPayment` transaction
     *
     * @example
     * ```ts
     * tx.add(client.paymentKit.call.processEphemeralPayment({ nonce, coinType, sender, amount, receiver }));
     * ```
     */
    this.processEphemeralPayment = (options) => {
      const { nonce, coinType, amount, receiver } = options;
      return processEphemeralPayment({
        arguments: {
          nonce,
          paymentAmount: amount,
          coin: coinWithBalance({
            type: coinType,
            balance: amount
          }),
          receiver
        },
        typeArguments: [coinType]
      });
    };
    /**
     * Creates a `createRegistry` transaction
     *
     * @example
     * ```ts
     * tx.add(client.paymentKit.call.createRegistry(registryName));
     * ```
     */
    this.createRegistry = ({ registryName }) => {
      return createRegistry({
        arguments: {
          namespace: __privateGet(this, _packageConfig).namespaceId,
          name: registryName
        }
      });
    };
    /**
     * Creates a `setConfigEpochExpirationDuration` transaction
     *
     * @example
     * ```ts
     * tx.add(client.paymentKit.call.setConfigEpochExpirationDuration({registryName, epochExpirationDuration, adminCapId}));
     * ```
     */
    this.setConfigEpochExpirationDuration = ({
      registryName,
      registryId,
      epochExpirationDuration,
      adminCapId
    }) => {
      const registryIdToUse = registryId ?? getRegistryIdFromName(registryName, __privateGet(this, _packageConfig).namespaceId);
      return setConfigEpochExpirationDuration({
        arguments: {
          registry: registryIdToUse,
          epochExpirationDuration,
          cap: adminCapId
        }
      });
    };
    /**
     * Creates a `setConfigRegistryManagedFunds` transaction
     *
     * @example
     * ```ts
     * tx.add(client.paymentKit.call.setConfigRegistryManagedFunds({registryName, registryManagedFunds, adminCapId}));
     * ```
     */
    this.setConfigRegistryManagedFunds = ({
      registryName,
      registryId,
      registryManagedFunds,
      adminCapId
    }) => {
      const registryIdToUse = registryId ?? getRegistryIdFromName(registryName, __privateGet(this, _packageConfig).namespaceId);
      return setConfigRegistryManagedFunds({
        arguments: {
          registry: registryIdToUse,
          registryManagedFunds,
          cap: adminCapId
        }
      });
    };
    /**
     * Creates a `withdrawFromRegistry` transaction
     *
     * @example
     * ```ts
     * tx.add(client.paymentKit.call.withdrawFromRegistry({coinType, registryName, adminCapId}));
     * ```
     */
    this.withdrawFromRegistry = ({
      coinType,
      registryName,
      registryId,
      adminCapId
    }) => {
      const registryIdToUse = registryId ?? getRegistryIdFromName(registryName, __privateGet(this, _packageConfig).namespaceId);
      return withdrawFromRegistry({
        arguments: {
          registry: registryIdToUse,
          cap: adminCapId
        },
        typeArguments: [coinType]
      });
    };
    /**
     * Creates a `deletePaymentRecord` transaction
     *
     * @example
     * ```ts
     * tx.add(client.paymentKit.call.deletePaymentRecord({coinType, nonce, amount, receiver, registryName}));
     * ```
     */
    this.deletePaymentRecord = ({
      coinType,
      nonce,
      amount,
      receiver,
      registryName,
      registryId
    }) => {
      const registryIdToUse = registryId ?? getRegistryIdFromName(registryName, __privateGet(this, _packageConfig).namespaceId);
      return deletePaymentRecord({
        arguments: {
          registry: registryIdToUse,
          paymentKey: createPaymentKey({
            arguments: {
              nonce,
              paymentAmount: amount,
              receiver
            },
            typeArguments: [coinType]
          })
        },
        typeArguments: [coinType]
      });
    };
    __privateSet(this, _packageConfig, options.packageConfig);
  }
}
_packageConfig = new WeakMap();
export {
  PaymentKitCalls
};
