// @mysten/payment-kit@0.0.17/dist/esm/client.js
var __typeError = (msg) => {
  throw TypeError(msg);
};
var __accessCheck = (obj, member, msg) => member.has(obj) || __typeError("Cannot " + msg);
var __privateGet = (obj, member, getter) => (__accessCheck(obj, member, "read from private field"), getter ? getter.call(obj) : member.get(obj));
var __privateAdd = (obj, member, value) => member.has(obj) ? __typeError("Cannot add the same private member more than once") : member instanceof WeakSet ? member.add(obj) : member.set(obj, value);
var __privateSet = (obj, member, value, setter) => (__accessCheck(obj, member, "write to private field"), setter ? setter.call(obj, value) : member.set(obj, value), value);
var _packageConfig, _client;
import { PaymentKitClientError } from "./error.js";
import {
  MAINNET_PAYMENT_KIT_PACKAGE_CONFIG,
  TESTNET_PAYMENT_KIT_PACKAGE_CONFIG
} from "./constants.js";
import { PaymentKey, PaymentRecord } from "./contracts/payment_kit/payment_kit.js";
import { normalizeStructTag } from "@mysten/sui/utils";
import { PaymentKitTransactions } from "./transactions.js";
import { PaymentKitCalls } from "./calls.js";
import { getRegistryIdFromName } from "./utils.js";
function paymentKit({ name = "paymentKit" } = {}) {
  return {
    name,
    register: (client) => {
      return new PaymentKitClient({ client });
    }
  };
}
class PaymentKitClient {
  constructor(options) {
    __privateAdd(this, _packageConfig);
    __privateAdd(this, _client);
    if (options.client) {
      __privateSet(this, _client, options.client);
    } else {
      throw new PaymentKitClientError("suiClient must be provided");
    }
    const network = options.client.network;
    switch (network) {
      case "testnet":
        __privateSet(this, _packageConfig, TESTNET_PAYMENT_KIT_PACKAGE_CONFIG);
        break;
      case "mainnet":
        __privateSet(this, _packageConfig, MAINNET_PAYMENT_KIT_PACKAGE_CONFIG);
        break;
      default:
        throw new PaymentKitClientError(`Unsupported network: ${network}`);
    }
    this.calls = new PaymentKitCalls({ packageConfig: __privateGet(this, _packageConfig) });
    this.tx = new PaymentKitTransactions({
      calls: this.calls
    });
  }
  /**
   * Query for a payment record in a registry.
   * Returns the payment record data if it exists, null otherwise.
   *
   * @example
   * ```ts
   * const paymentRecord = await client.getPaymentRecord({ registry, nonce, amount, receiver, coinType });
   * ```
   */
  async getPaymentRecord(options) {
    const { coinType, registryId, registryName, nonce, amount, receiver } = options;
    const normalizedCoinType = normalizeStructTag(coinType);
    const paymentKeyType = `${PaymentKey.name}<${normalizedCoinType}>`;
    const registryIdToUse = registryId ?? getRegistryIdFromName(registryName, __privateGet(this, _packageConfig).namespaceId);
    const result = await __privateGet(this, _client).core.getDynamicField({
      parentId: registryIdToUse,
      name: {
        type: paymentKeyType,
        bcs: PaymentKey.serialize({
          nonce,
          payment_amount: amount,
          receiver
        }).toBytes()
      }
    });
    if (!result?.dynamicField) {
      return null;
    }
    const decoded = PaymentRecord.parse(result.dynamicField.value.bcs);
    return {
      key: result.dynamicField.id,
      paymentTransactionDigest: result.dynamicField.previousTransaction,
      epochAtTimeOfRecord: decoded.epoch_at_time_of_record
    };
  }
  /**
   * Get the registry object id from a registry name.
   * Returns the derived registry id.
   *
   * @example
   * ```ts
   * const registryId = await client.getRegistryIdFromName("my-registry");
   * ```
   */
  getRegistryIdFromName(registryName) {
    return getRegistryIdFromName(registryName, __privateGet(this, _packageConfig).namespaceId);
  }
}
_packageConfig = new WeakMap();
_client = new WeakMap();
export {
  PaymentKitClient,
  paymentKit
};
