// @mysten/payment-kit@0.0.17/dist/esm/transactions.js
var __typeError = (msg) => {
  throw TypeError(msg);
};
var __accessCheck = (obj, member, msg) => member.has(obj) || __typeError("Cannot " + msg);
var __privateGet = (obj, member, getter) => (__accessCheck(obj, member, "read from private field"), getter ? getter.call(obj) : member.get(obj));
var __privateAdd = (obj, member, value) => member.has(obj) ? __typeError("Cannot add the same private member more than once") : member instanceof WeakSet ? member.add(obj) : member.set(obj, value);
var __privateSet = (obj, member, value, setter) => (__accessCheck(obj, member, "write to private field"), setter ? setter.call(obj, value) : member.set(obj, value), value);
var _calls;
import { Transaction } from "@mysten/sui/transactions";
class PaymentKitTransactions {
  constructor(options) {
    __privateAdd(this, _calls);
    __privateSet(this, _calls, options.calls);
  }
  /**
   * Creates a `processRegistryPayment` transaction
   *
   * @example
   * ```ts
   * const tx = client.paymentKit.tx.processRegistryPayment({ nonce, coinType, sender, amount, receiver, registryName });
   * ```
   */
  processRegistryPayment(options) {
    const tx = new Transaction();
    tx.add(__privateGet(this, _calls).processRegistryPayment(options));
    return tx;
  }
  /**
   * Creates a `processEphemeralPayment` transaction
   *
   * @example
   * ```ts
   * const tx = client.paymentKit.tx.processEphemeralPayment({ nonce, coinType, sender, amount, receiver });
   * ```
   */
  processEphemeralPayment(options) {
    const tx = new Transaction();
    tx.add(__privateGet(this, _calls).processEphemeralPayment(options));
    return tx;
  }
  /**
   * Creates a `createRegistry` transaction
   *
   * @example
   * ```ts
   * const tx = client.paymentKit.tx.createRegistry({ registryName });
   * ```
   */
  createRegistry(options) {
    const tx = new Transaction();
    tx.add(__privateGet(this, _calls).createRegistry(options));
    return tx;
  }
  /**
   * Creates a `setConfigEpochExpirationDuration` transaction
   *
   * @example
   * ```ts
   * const tx = client.paymentKit.tx.setConfigEpochExpirationDuration({ registryName, epochExpirationDuration, adminCapId });
   * ```
   */
  setConfigEpochExpirationDuration(options) {
    const tx = new Transaction();
    tx.add(__privateGet(this, _calls).setConfigEpochExpirationDuration(options));
    return tx;
  }
  /**
   * Creates a `setConfigRegistryManagedFunds` transaction
   *
   * @example
   * ```ts
   * const tx = client.paymentKit.tx.setConfigRegistryManagedFunds({ registryName, registryManagedFunds, adminCapId });
   * ```
   */
  setConfigRegistryManagedFunds(options) {
    const tx = new Transaction();
    tx.add(__privateGet(this, _calls).setConfigRegistryManagedFunds(options));
    return tx;
  }
  /**
   * Creates a `withdrawFromRegistry` transaction
   *
   * @example
   * ```ts
   * const tx = client.paymentKit.tx.withdrawFromRegistry({ coinType, registryName, adminCapId });
   * ```
   */
  withdrawFromRegistry(options) {
    const tx = new Transaction();
    tx.add(__privateGet(this, _calls).withdrawFromRegistry(options));
    return tx;
  }
  /**
   * Creates a `deletePaymentRecord` transaction
   *
   * @example
   * ```ts
   * const tx = client.paymentKit.tx.deletePaymentRecord({ coinType, nonce, amount, receiver, registryName });
   * ```
   */
  deletePaymentRecord(options) {
    const tx = new Transaction();
    tx.add(__privateGet(this, _calls).deletePaymentRecord(options));
    return tx;
  }
}
_calls = new WeakMap();
export {
  PaymentKitTransactions
};
