// @mysten/payment-kit@0.0.17/dist/esm/uri.js
import { isValidNamedType, isValidSuiAddress, isValidSuiObjectId } from "@mysten/sui/utils";
import { PaymentKitUriError } from "./error.js";
import { SUI_PAYMENT_KIT_PROTOCOL } from "./constants.js";
const isValidNonce = (nonce) => {
  return nonce.length <= 36;
};
const isValidAmount = (amount) => {
  return amount > 0n;
};
const isValidCoinType = (coinType) => {
  return isValidNamedType(coinType);
};
const createPaymentTransactionUri = (params) => {
  const { receiverAddress, amount, coinType, nonce, registryId, registryName } = params;
  const uri = new URL(SUI_PAYMENT_KIT_PROTOCOL);
  if (isValidSuiAddress(receiverAddress)) {
    uri.searchParams.append("receiver", receiverAddress);
  } else {
    throw new PaymentKitUriError("Invalid Sui address");
  }
  if (isValidAmount(amount)) {
    uri.searchParams.append("amount", amount.toString());
  } else {
    throw new PaymentKitUriError("Amount must be a positive numeric string");
  }
  if (isValidCoinType(coinType)) {
    uri.searchParams.append("coinType", coinType);
  } else {
    throw new PaymentKitUriError("Invalid Coin Type");
  }
  if (isValidNonce(nonce)) {
    uri.searchParams.append("nonce", nonce);
  } else {
    throw new PaymentKitUriError("Nonce length exceeds maximum of 36 characters");
  }
  if (registryId) {
    if (isValidSuiObjectId(registryId)) {
      uri.searchParams.append("registry", registryId);
    } else {
      throw new PaymentKitUriError("Invalid Sui Object Id for Registry Id");
    }
  }
  if (registryName) {
    uri.searchParams.append("registry", registryName);
  }
  if (params.label) {
    uri.searchParams.append("label", params.label);
  }
  if (params.message) {
    uri.searchParams.append("message", params.message);
  }
  if (params.iconUrl) {
    uri.searchParams.append("iconUrl", params.iconUrl);
  }
  return uri.toString();
};
const parsePaymentTransactionUri = (uri) => {
  if (!uri.startsWith(SUI_PAYMENT_KIT_PROTOCOL + "?")) {
    throw new PaymentKitUriError("Invalid URI: Must start with sui:pay?");
  }
  const url = new URL(uri);
  const params = url.searchParams;
  const receiver = params.get("receiver");
  const amount = params.get("amount");
  const coinType = params.get("coinType");
  const nonce = params.get("nonce");
  if (!receiver || !amount || !coinType || !nonce) {
    throw new PaymentKitUriError("Invalid URI: Missing required parameters");
  }
  if (!isValidSuiAddress(receiver)) {
    throw new PaymentKitUriError("Invalid URI: Receiver address is not valid");
  }
  if (!isValidCoinType(coinType)) {
    throw new PaymentKitUriError("Invalid URI: Coin Type is not valid");
  }
  if (!isValidNonce(nonce)) {
    throw new PaymentKitUriError("Invalid URI: Nonce length exceeds maximum of 36 characters");
  }
  const bigIntAmount = BigInt(amount);
  if (!isValidAmount(bigIntAmount)) {
    throw new PaymentKitUriError("Invalid URI: Amount must be a positive number");
  }
  const registry = params.get("registry") ?? void 0;
  let registryId;
  let registryName;
  if (registry) {
    if (isValidSuiObjectId(registry)) {
      registryId = registry;
    } else {
      registryName = registry;
    }
  }
  const baseParams = {
    receiverAddress: receiver,
    amount: bigIntAmount,
    coinType,
    nonce,
    label: params.get("label") ?? void 0,
    message: params.get("message") ?? void 0,
    iconUrl: params.get("icon") ?? void 0
  };
  if (registryId) {
    return { ...baseParams, registryId };
  }
  return { ...baseParams, registryName };
};
export {
  createPaymentTransactionUri,
  parsePaymentTransactionUri
};
