// @mysten/payment-kit@0.0.17/dist/cjs/contracts/payment_kit/payment_kit.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var payment_kit_exports = {};
__export(payment_kit_exports, {
  BalanceKey: () => BalanceKey,
  Namespace: () => Namespace,
  PaymentKey: () => PaymentKey,
  PaymentReceipt: () => PaymentReceipt,
  PaymentRecord: () => PaymentRecord,
  PaymentRegistry: () => PaymentRegistry,
  PaymentType: () => PaymentType,
  RegistryAdminCap: () => RegistryAdminCap,
  createPaymentKey: () => createPaymentKey,
  createRegistry: () => createRegistry,
  deletePaymentRecord: () => deletePaymentRecord,
  processEphemeralPayment: () => processEphemeralPayment,
  processRegistryPayment: () => processRegistryPayment,
  setConfigEpochExpirationDuration: () => setConfigEpochExpirationDuration,
  setConfigRegistryManagedFunds: () => setConfigRegistryManagedFunds,
  share: () => share,
  withdrawFromRegistry: () => withdrawFromRegistry
});
module.exports = __toCommonJS(payment_kit_exports);
var import_utils = require("../utils/index.js");
var import_bcs = require("@mysten/sui/bcs");
var object = __toESM(require("./deps/sui/object.js"));
var vec_map = __toESM(require("./deps/sui/vec_map.js"));
var config = __toESM(require("./config.js"));
const $moduleName = "@mysten/payment-kit::payment_kit";
const Namespace = new import_utils.MoveStruct({
  name: `${$moduleName}::Namespace`,
  fields: {
    id: object.UID
  }
});
const PaymentRegistry = new import_utils.MoveStruct({
  name: `${$moduleName}::PaymentRegistry`,
  fields: {
    id: object.UID,
    cap_id: import_bcs.bcs.Address,
    config: vec_map.VecMap(import_bcs.bcs.string(), config.Value),
    version: import_bcs.bcs.u16()
  }
});
const RegistryAdminCap = new import_utils.MoveStruct({
  name: `${$moduleName}::RegistryAdminCap`,
  fields: {
    id: object.UID,
    registry_id: import_bcs.bcs.Address
  }
});
const PaymentType = new import_utils.MoveEnum({
  name: `${$moduleName}::PaymentType`,
  fields: {
    Ephemeral: null,
    Registry: import_bcs.bcs.Address
  }
});
const PaymentReceipt = new import_utils.MoveStruct({
  name: `${$moduleName}::PaymentReceipt`,
  fields: {
    payment_type: PaymentType,
    nonce: import_bcs.bcs.string(),
    payment_amount: import_bcs.bcs.u64(),
    receiver: import_bcs.bcs.Address,
    coin_type: import_bcs.bcs.string(),
    timestamp_ms: import_bcs.bcs.u64()
  }
});
const PaymentKey = new import_utils.MoveStruct({
  name: `${$moduleName}::PaymentKey`,
  fields: {
    nonce: import_bcs.bcs.string(),
    payment_amount: import_bcs.bcs.u64(),
    receiver: import_bcs.bcs.Address
  }
});
const PaymentRecord = new import_utils.MoveStruct({
  name: `${$moduleName}::PaymentRecord`,
  fields: {
    epoch_at_time_of_record: import_bcs.bcs.u64()
  }
});
const BalanceKey = new import_utils.MoveTuple({
  name: `${$moduleName}::BalanceKey`,
  fields: [import_bcs.bcs.bool()]
});
function createRegistry(options) {
  const packageAddress = options.package ?? "@mysten/payment-kit";
  const argumentsTypes = [
    `${packageAddress}::payment_kit::Namespace`,
    "0x0000000000000000000000000000000000000000000000000000000000000001::ascii::String"
  ];
  const parameterNames = ["namespace", "name"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "payment_kit",
    function: "create_registry",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames)
  });
}
function processEphemeralPayment(options) {
  const packageAddress = options.package ?? "@mysten/payment-kit";
  const argumentsTypes = [
    "0x0000000000000000000000000000000000000000000000000000000000000001::ascii::String",
    "u64",
    `0x0000000000000000000000000000000000000000000000000000000000000002::coin::Coin<${options.typeArguments[0]}>`,
    "address",
    "0x0000000000000000000000000000000000000000000000000000000000000002::clock::Clock"
  ];
  const parameterNames = ["nonce", "paymentAmount", "coin", "receiver"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "payment_kit",
    function: "process_ephemeral_payment",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames),
    typeArguments: options.typeArguments
  });
}
function processRegistryPayment(options) {
  const packageAddress = options.package ?? "@mysten/payment-kit";
  const argumentsTypes = [
    `${packageAddress}::payment_kit::PaymentRegistry`,
    "0x0000000000000000000000000000000000000000000000000000000000000001::ascii::String",
    "u64",
    `0x0000000000000000000000000000000000000000000000000000000000000002::coin::Coin<${options.typeArguments[0]}>`,
    "0x0000000000000000000000000000000000000000000000000000000000000001::option::Option<address>",
    "0x0000000000000000000000000000000000000000000000000000000000000002::clock::Clock"
  ];
  const parameterNames = ["registry", "nonce", "paymentAmount", "coin", "receiver"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "payment_kit",
    function: "process_registry_payment",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames),
    typeArguments: options.typeArguments
  });
}
function withdrawFromRegistry(options) {
  const packageAddress = options.package ?? "@mysten/payment-kit";
  const argumentsTypes = [
    `${packageAddress}::payment_kit::PaymentRegistry`,
    `${packageAddress}::payment_kit::RegistryAdminCap`
  ];
  const parameterNames = ["registry", "cap"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "payment_kit",
    function: "withdraw_from_registry",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames),
    typeArguments: options.typeArguments
  });
}
function deletePaymentRecord(options) {
  const packageAddress = options.package ?? "@mysten/payment-kit";
  const argumentsTypes = [
    `${packageAddress}::payment_kit::PaymentRegistry`,
    `${packageAddress}::payment_kit::PaymentKey<${options.typeArguments[0]}>`
  ];
  const parameterNames = ["registry", "paymentKey"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "payment_kit",
    function: "delete_payment_record",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames),
    typeArguments: options.typeArguments
  });
}
function createPaymentKey(options) {
  const packageAddress = options.package ?? "@mysten/payment-kit";
  const argumentsTypes = [
    "0x0000000000000000000000000000000000000000000000000000000000000001::ascii::String",
    "u64",
    "address"
  ];
  const parameterNames = ["nonce", "paymentAmount", "receiver"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "payment_kit",
    function: "create_payment_key",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames),
    typeArguments: options.typeArguments
  });
}
function setConfigEpochExpirationDuration(options) {
  const packageAddress = options.package ?? "@mysten/payment-kit";
  const argumentsTypes = [
    `${packageAddress}::payment_kit::PaymentRegistry`,
    `${packageAddress}::payment_kit::RegistryAdminCap`,
    "u64"
  ];
  const parameterNames = ["registry", "cap", "epochExpirationDuration"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "payment_kit",
    function: "set_config_epoch_expiration_duration",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames)
  });
}
function setConfigRegistryManagedFunds(options) {
  const packageAddress = options.package ?? "@mysten/payment-kit";
  const argumentsTypes = [
    `${packageAddress}::payment_kit::PaymentRegistry`,
    `${packageAddress}::payment_kit::RegistryAdminCap`,
    "bool"
  ];
  const parameterNames = ["registry", "cap", "registryManagedFunds"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "payment_kit",
    function: "set_config_registry_managed_funds",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames)
  });
}
function share(options) {
  const packageAddress = options.package ?? "@mysten/payment-kit";
  const argumentsTypes = [`${packageAddress}::payment_kit::PaymentRegistry`];
  const parameterNames = ["registry"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "payment_kit",
    function: "share",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames)
  });
}
