// @mysten/payment-kit@0.0.17/dist/cjs/calls.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __typeError = (msg) => {
  throw TypeError(msg);
};
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var __accessCheck = (obj, member, msg) => member.has(obj) || __typeError("Cannot " + msg);
var __privateGet = (obj, member, getter) => (__accessCheck(obj, member, "read from private field"), getter ? getter.call(obj) : member.get(obj));
var __privateAdd = (obj, member, value) => member.has(obj) ? __typeError("Cannot add the same private member more than once") : member instanceof WeakSet ? member.add(obj) : member.set(obj, value);
var __privateSet = (obj, member, value, setter) => (__accessCheck(obj, member, "write to private field"), setter ? setter.call(obj, value) : member.set(obj, value), value);
var calls_exports = {};
__export(calls_exports, {
  PaymentKitCalls: () => PaymentKitCalls
});
module.exports = __toCommonJS(calls_exports);
var import_transactions = require("@mysten/sui/transactions");
var import_payment_kit = require("./contracts/payment_kit/payment_kit.js");
var import_utils = require("./utils.js");
var _packageConfig;
class PaymentKitCalls {
  constructor(options) {
    __privateAdd(this, _packageConfig);
    /**
     * Creates a `processRegistryPayment` transaction
     *
     * @example
     * ```ts
     * tx.add(lient.paymentKit.call.processRegistryPayment({ nonce, coinType, sender, amount, receiver, registryName }));
     * ```
     */
    this.processRegistryPayment = (options) => {
      const { nonce, coinType, amount, receiver, registryName, registryId } = options;
      const registryIdToUse = registryId ?? (0, import_utils.getRegistryIdFromName)(registryName, __privateGet(this, _packageConfig).namespaceId);
      return (0, import_payment_kit.processRegistryPayment)({
        arguments: {
          registry: registryIdToUse,
          nonce,
          paymentAmount: amount,
          coin: (0, import_transactions.coinWithBalance)({
            type: coinType,
            balance: amount
          }),
          receiver
        },
        typeArguments: [coinType]
      });
    };
    /**
     * Creates a `processRegistryPayment` transaction
     *
     * @example
     * ```ts
     * tx.add(client.paymentKit.call.processEphemeralPayment({ nonce, coinType, sender, amount, receiver }));
     * ```
     */
    this.processEphemeralPayment = (options) => {
      const { nonce, coinType, amount, receiver } = options;
      return (0, import_payment_kit.processEphemeralPayment)({
        arguments: {
          nonce,
          paymentAmount: amount,
          coin: (0, import_transactions.coinWithBalance)({
            type: coinType,
            balance: amount
          }),
          receiver
        },
        typeArguments: [coinType]
      });
    };
    /**
     * Creates a `createRegistry` transaction
     *
     * @example
     * ```ts
     * tx.add(client.paymentKit.call.createRegistry(registryName));
     * ```
     */
    this.createRegistry = ({ registryName }) => {
      return (0, import_payment_kit.createRegistry)({
        arguments: {
          namespace: __privateGet(this, _packageConfig).namespaceId,
          name: registryName
        }
      });
    };
    /**
     * Creates a `setConfigEpochExpirationDuration` transaction
     *
     * @example
     * ```ts
     * tx.add(client.paymentKit.call.setConfigEpochExpirationDuration({registryName, epochExpirationDuration, adminCapId}));
     * ```
     */
    this.setConfigEpochExpirationDuration = ({
      registryName,
      registryId,
      epochExpirationDuration,
      adminCapId
    }) => {
      const registryIdToUse = registryId ?? (0, import_utils.getRegistryIdFromName)(registryName, __privateGet(this, _packageConfig).namespaceId);
      return (0, import_payment_kit.setConfigEpochExpirationDuration)({
        arguments: {
          registry: registryIdToUse,
          epochExpirationDuration,
          cap: adminCapId
        }
      });
    };
    /**
     * Creates a `setConfigRegistryManagedFunds` transaction
     *
     * @example
     * ```ts
     * tx.add(client.paymentKit.call.setConfigRegistryManagedFunds({registryName, registryManagedFunds, adminCapId}));
     * ```
     */
    this.setConfigRegistryManagedFunds = ({
      registryName,
      registryId,
      registryManagedFunds,
      adminCapId
    }) => {
      const registryIdToUse = registryId ?? (0, import_utils.getRegistryIdFromName)(registryName, __privateGet(this, _packageConfig).namespaceId);
      return (0, import_payment_kit.setConfigRegistryManagedFunds)({
        arguments: {
          registry: registryIdToUse,
          registryManagedFunds,
          cap: adminCapId
        }
      });
    };
    /**
     * Creates a `withdrawFromRegistry` transaction
     *
     * @example
     * ```ts
     * tx.add(client.paymentKit.call.withdrawFromRegistry({coinType, registryName, adminCapId}));
     * ```
     */
    this.withdrawFromRegistry = ({
      coinType,
      registryName,
      registryId,
      adminCapId
    }) => {
      const registryIdToUse = registryId ?? (0, import_utils.getRegistryIdFromName)(registryName, __privateGet(this, _packageConfig).namespaceId);
      return (0, import_payment_kit.withdrawFromRegistry)({
        arguments: {
          registry: registryIdToUse,
          cap: adminCapId
        },
        typeArguments: [coinType]
      });
    };
    /**
     * Creates a `deletePaymentRecord` transaction
     *
     * @example
     * ```ts
     * tx.add(client.paymentKit.call.deletePaymentRecord({coinType, nonce, amount, receiver, registryName}));
     * ```
     */
    this.deletePaymentRecord = ({
      coinType,
      nonce,
      amount,
      receiver,
      registryName,
      registryId
    }) => {
      const registryIdToUse = registryId ?? (0, import_utils.getRegistryIdFromName)(registryName, __privateGet(this, _packageConfig).namespaceId);
      return (0, import_payment_kit.deletePaymentRecord)({
        arguments: {
          registry: registryIdToUse,
          paymentKey: (0, import_payment_kit.createPaymentKey)({
            arguments: {
              nonce,
              paymentAmount: amount,
              receiver
            },
            typeArguments: [coinType]
          })
        },
        typeArguments: [coinType]
      });
    };
    __privateSet(this, _packageConfig, options.packageConfig);
  }
}
_packageConfig = new WeakMap();
