// @mysten/payment-kit@0.0.17/dist/cjs/client.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __typeError = (msg) => {
  throw TypeError(msg);
};
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var __accessCheck = (obj, member, msg) => member.has(obj) || __typeError("Cannot " + msg);
var __privateGet = (obj, member, getter) => (__accessCheck(obj, member, "read from private field"), getter ? getter.call(obj) : member.get(obj));
var __privateAdd = (obj, member, value) => member.has(obj) ? __typeError("Cannot add the same private member more than once") : member instanceof WeakSet ? member.add(obj) : member.set(obj, value);
var __privateSet = (obj, member, value, setter) => (__accessCheck(obj, member, "write to private field"), setter ? setter.call(obj, value) : member.set(obj, value), value);
var client_exports = {};
__export(client_exports, {
  PaymentKitClient: () => PaymentKitClient,
  paymentKit: () => paymentKit
});
module.exports = __toCommonJS(client_exports);
var import_error = require("./error.js");
var import_constants = require("./constants.js");
var import_payment_kit = require("./contracts/payment_kit/payment_kit.js");
var import_utils = require("@mysten/sui/utils");
var import_transactions = require("./transactions.js");
var import_calls = require("./calls.js");
var import_utils2 = require("./utils.js");
var _packageConfig, _client;
function paymentKit({ name = "paymentKit" } = {}) {
  return {
    name,
    register: (client) => {
      return new PaymentKitClient({ client });
    }
  };
}
class PaymentKitClient {
  constructor(options) {
    __privateAdd(this, _packageConfig);
    __privateAdd(this, _client);
    if (options.client) {
      __privateSet(this, _client, options.client);
    } else {
      throw new import_error.PaymentKitClientError("suiClient must be provided");
    }
    const network = options.client.network;
    switch (network) {
      case "testnet":
        __privateSet(this, _packageConfig, import_constants.TESTNET_PAYMENT_KIT_PACKAGE_CONFIG);
        break;
      case "mainnet":
        __privateSet(this, _packageConfig, import_constants.MAINNET_PAYMENT_KIT_PACKAGE_CONFIG);
        break;
      default:
        throw new import_error.PaymentKitClientError(`Unsupported network: ${network}`);
    }
    this.calls = new import_calls.PaymentKitCalls({ packageConfig: __privateGet(this, _packageConfig) });
    this.tx = new import_transactions.PaymentKitTransactions({
      calls: this.calls
    });
  }
  /**
   * Query for a payment record in a registry.
   * Returns the payment record data if it exists, null otherwise.
   *
   * @example
   * ```ts
   * const paymentRecord = await client.getPaymentRecord({ registry, nonce, amount, receiver, coinType });
   * ```
   */
  async getPaymentRecord(options) {
    const { coinType, registryId, registryName, nonce, amount, receiver } = options;
    const normalizedCoinType = (0, import_utils.normalizeStructTag)(coinType);
    const paymentKeyType = `${import_payment_kit.PaymentKey.name}<${normalizedCoinType}>`;
    const registryIdToUse = registryId ?? (0, import_utils2.getRegistryIdFromName)(registryName, __privateGet(this, _packageConfig).namespaceId);
    const result = await __privateGet(this, _client).core.getDynamicField({
      parentId: registryIdToUse,
      name: {
        type: paymentKeyType,
        bcs: import_payment_kit.PaymentKey.serialize({
          nonce,
          payment_amount: amount,
          receiver
        }).toBytes()
      }
    });
    if (!result?.dynamicField) {
      return null;
    }
    const decoded = import_payment_kit.PaymentRecord.parse(result.dynamicField.value.bcs);
    return {
      key: result.dynamicField.id,
      paymentTransactionDigest: result.dynamicField.previousTransaction,
      epochAtTimeOfRecord: decoded.epoch_at_time_of_record
    };
  }
  /**
   * Get the registry object id from a registry name.
   * Returns the derived registry id.
   *
   * @example
   * ```ts
   * const registryId = await client.getRegistryIdFromName("my-registry");
   * ```
   */
  getRegistryIdFromName(registryName) {
    return (0, import_utils2.getRegistryIdFromName)(registryName, __privateGet(this, _packageConfig).namespaceId);
  }
}
_packageConfig = new WeakMap();
_client = new WeakMap();
