// @mysten/payment-kit@0.0.17/dist/cjs/constants.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var constants_exports = {};
__export(constants_exports, {
  DEFAULT_REGISTRY_NAME: () => DEFAULT_REGISTRY_NAME,
  MAINNET_PAYMENT_KIT_PACKAGE_CONFIG: () => MAINNET_PAYMENT_KIT_PACKAGE_CONFIG,
  SUI_COIN_TYPE: () => SUI_COIN_TYPE,
  SUI_PAYMENT_KIT_PROTOCOL: () => SUI_PAYMENT_KIT_PROTOCOL,
  TESTNET_PAYMENT_KIT_PACKAGE_CONFIG: () => TESTNET_PAYMENT_KIT_PACKAGE_CONFIG
});
module.exports = __toCommonJS(constants_exports);
var import_utils = require("@mysten/sui/utils");
const TESTNET_PAYMENT_KIT_PACKAGE_CONFIG = {
  packageId: "0x7e069abe383e80d32f2aec17b3793da82aabc8c2edf84abbf68dd7b719e71497",
  namespaceId: "0xa5016862fdccba7cc576b56cc5a391eda6775200aaa03a6b3c97d512312878db"
};
const MAINNET_PAYMENT_KIT_PACKAGE_CONFIG = {
  packageId: "0xbc126f1535fba7d641cb9150ad9eae93b104972586ba20f3c60bfe0e53b69bc6",
  namespaceId: "0xccd3e4c7802921991cd9ce488c4ca0b51334ba75483702744242284ccf3ae7c2"
};
const SUI_COIN_TYPE = (0, import_utils.normalizeStructTag)(import_utils.SUI_TYPE_ARG);
const DEFAULT_REGISTRY_NAME = "default-payment-registry";
const SUI_PAYMENT_KIT_PROTOCOL = "sui:pay";
