// @mysten/payment-kit@0.0.17/dist/cjs/index.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var index_exports = {};
__export(index_exports, {
  DEFAULT_REGISTRY_NAME: () => import_constants.DEFAULT_REGISTRY_NAME,
  PaymentKitClient: () => import_client.PaymentKitClient,
  PaymentKitClientError: () => import_error.PaymentKitClientError,
  PaymentKitUriError: () => import_error.PaymentKitUriError,
  SUI_PAYMENT_KIT_PROTOCOL: () => import_constants.SUI_PAYMENT_KIT_PROTOCOL,
  createPaymentTransactionUri: () => import_uri.createPaymentTransactionUri,
  parsePaymentTransactionUri: () => import_uri.parsePaymentTransactionUri,
  paymentKit: () => import_client.paymentKit
});
module.exports = __toCommonJS(index_exports);
var import_client = require("./client.js");
var import_error = require("./error.js");
var import_constants = require("./constants.js");
var import_uri = require("./uri.js");
