// @mysten/payment-kit@0.0.17/dist/esm/contracts/payment_kit/payment_kit.js
import { MoveStruct, MoveEnum, MoveTuple, normalizeMoveArguments } from "../utils/index.js";
import { bcs } from "@mysten/sui/bcs";
import * as object from "./deps/sui/object.js";
import * as vec_map from "./deps/sui/vec_map.js";
import * as config from "./config.js";
const $moduleName = "@mysten/payment-kit::payment_kit";
const Namespace = new MoveStruct({
  name: `${$moduleName}::Namespace`,
  fields: {
    id: object.UID
  }
});
const PaymentRegistry = new MoveStruct({
  name: `${$moduleName}::PaymentRegistry`,
  fields: {
    id: object.UID,
    cap_id: bcs.Address,
    config: vec_map.VecMap(bcs.string(), config.Value),
    version: bcs.u16()
  }
});
const RegistryAdminCap = new MoveStruct({
  name: `${$moduleName}::RegistryAdminCap`,
  fields: {
    id: object.UID,
    registry_id: bcs.Address
  }
});
const PaymentType = new MoveEnum({
  name: `${$moduleName}::PaymentType`,
  fields: {
    Ephemeral: null,
    Registry: bcs.Address
  }
});
const PaymentReceipt = new MoveStruct({
  name: `${$moduleName}::PaymentReceipt`,
  fields: {
    payment_type: PaymentType,
    nonce: bcs.string(),
    payment_amount: bcs.u64(),
    receiver: bcs.Address,
    coin_type: bcs.string(),
    timestamp_ms: bcs.u64()
  }
});
const PaymentKey = new MoveStruct({
  name: `${$moduleName}::PaymentKey`,
  fields: {
    nonce: bcs.string(),
    payment_amount: bcs.u64(),
    receiver: bcs.Address
  }
});
const PaymentRecord = new MoveStruct({
  name: `${$moduleName}::PaymentRecord`,
  fields: {
    epoch_at_time_of_record: bcs.u64()
  }
});
const BalanceKey = new MoveTuple({
  name: `${$moduleName}::BalanceKey`,
  fields: [bcs.bool()]
});
function createRegistry(options) {
  const packageAddress = options.package ?? "@mysten/payment-kit";
  const argumentsTypes = [
    `${packageAddress}::payment_kit::Namespace`,
    "0x0000000000000000000000000000000000000000000000000000000000000001::ascii::String"
  ];
  const parameterNames = ["namespace", "name"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "payment_kit",
    function: "create_registry",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function processEphemeralPayment(options) {
  const packageAddress = options.package ?? "@mysten/payment-kit";
  const argumentsTypes = [
    "0x0000000000000000000000000000000000000000000000000000000000000001::ascii::String",
    "u64",
    `0x0000000000000000000000000000000000000000000000000000000000000002::coin::Coin<${options.typeArguments[0]}>`,
    "address",
    "0x0000000000000000000000000000000000000000000000000000000000000002::clock::Clock"
  ];
  const parameterNames = ["nonce", "paymentAmount", "coin", "receiver"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "payment_kit",
    function: "process_ephemeral_payment",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames),
    typeArguments: options.typeArguments
  });
}
function processRegistryPayment(options) {
  const packageAddress = options.package ?? "@mysten/payment-kit";
  const argumentsTypes = [
    `${packageAddress}::payment_kit::PaymentRegistry`,
    "0x0000000000000000000000000000000000000000000000000000000000000001::ascii::String",
    "u64",
    `0x0000000000000000000000000000000000000000000000000000000000000002::coin::Coin<${options.typeArguments[0]}>`,
    "0x0000000000000000000000000000000000000000000000000000000000000001::option::Option<address>",
    "0x0000000000000000000000000000000000000000000000000000000000000002::clock::Clock"
  ];
  const parameterNames = ["registry", "nonce", "paymentAmount", "coin", "receiver"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "payment_kit",
    function: "process_registry_payment",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames),
    typeArguments: options.typeArguments
  });
}
function withdrawFromRegistry(options) {
  const packageAddress = options.package ?? "@mysten/payment-kit";
  const argumentsTypes = [
    `${packageAddress}::payment_kit::PaymentRegistry`,
    `${packageAddress}::payment_kit::RegistryAdminCap`
  ];
  const parameterNames = ["registry", "cap"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "payment_kit",
    function: "withdraw_from_registry",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames),
    typeArguments: options.typeArguments
  });
}
function deletePaymentRecord(options) {
  const packageAddress = options.package ?? "@mysten/payment-kit";
  const argumentsTypes = [
    `${packageAddress}::payment_kit::PaymentRegistry`,
    `${packageAddress}::payment_kit::PaymentKey<${options.typeArguments[0]}>`
  ];
  const parameterNames = ["registry", "paymentKey"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "payment_kit",
    function: "delete_payment_record",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames),
    typeArguments: options.typeArguments
  });
}
function createPaymentKey(options) {
  const packageAddress = options.package ?? "@mysten/payment-kit";
  const argumentsTypes = [
    "0x0000000000000000000000000000000000000000000000000000000000000001::ascii::String",
    "u64",
    "address"
  ];
  const parameterNames = ["nonce", "paymentAmount", "receiver"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "payment_kit",
    function: "create_payment_key",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames),
    typeArguments: options.typeArguments
  });
}
function setConfigEpochExpirationDuration(options) {
  const packageAddress = options.package ?? "@mysten/payment-kit";
  const argumentsTypes = [
    `${packageAddress}::payment_kit::PaymentRegistry`,
    `${packageAddress}::payment_kit::RegistryAdminCap`,
    "u64"
  ];
  const parameterNames = ["registry", "cap", "epochExpirationDuration"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "payment_kit",
    function: "set_config_epoch_expiration_duration",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function setConfigRegistryManagedFunds(options) {
  const packageAddress = options.package ?? "@mysten/payment-kit";
  const argumentsTypes = [
    `${packageAddress}::payment_kit::PaymentRegistry`,
    `${packageAddress}::payment_kit::RegistryAdminCap`,
    "bool"
  ];
  const parameterNames = ["registry", "cap", "registryManagedFunds"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "payment_kit",
    function: "set_config_registry_managed_funds",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function share(options) {
  const packageAddress = options.package ?? "@mysten/payment-kit";
  const argumentsTypes = [`${packageAddress}::payment_kit::PaymentRegistry`];
  const parameterNames = ["registry"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "payment_kit",
    function: "share",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
export {
  BalanceKey,
  Namespace,
  PaymentKey,
  PaymentReceipt,
  PaymentRecord,
  PaymentRegistry,
  PaymentType,
  RegistryAdminCap,
  createPaymentKey,
  createRegistry,
  deletePaymentRecord,
  processEphemeralPayment,
  processRegistryPayment,
  setConfigEpochExpirationDuration,
  setConfigRegistryManagedFunds,
  share,
  withdrawFromRegistry
};
