// @mysten/kiosk@0.14.4/dist/cjs/bcs.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var bcs_exports = {};
__export(bcs_exports, {
  KioskPurchaseCap: () => KioskPurchaseCap,
  KioskType: () => KioskType,
  TransferPolicyCreatedEvent: () => TransferPolicyCreatedEvent,
  TransferPolicyType: () => TransferPolicyType
});
module.exports = __toCommonJS(bcs_exports);
var import_bcs = require("@mysten/sui/bcs");
var import_types = require("./types/index.js");
const KioskType = import_bcs.bcs.struct(import_types.KIOSK_TYPE, {
  id: import_bcs.bcs.Address,
  profits: import_bcs.bcs.u64(),
  owner: import_bcs.bcs.Address,
  itemCount: import_bcs.bcs.u32(),
  allowExtensions: import_bcs.bcs.bool()
});
const KioskPurchaseCap = import_bcs.bcs.struct(import_types.KIOSK_PURCHASE_CAP, {
  id: import_bcs.bcs.Address,
  kioskId: import_bcs.bcs.Address,
  itemId: import_bcs.bcs.Address,
  minPrice: import_bcs.bcs.u64()
});
const TransferPolicyCreatedEvent = import_bcs.bcs.struct(import_types.TRANSFER_POLICY_CREATED_EVENT, {
  id: import_bcs.bcs.Address
});
const TransferPolicyType = import_bcs.bcs.struct(import_types.TRANSFER_POLICY_TYPE, {
  id: import_bcs.bcs.Address,
  balance: import_bcs.bcs.u64(),
  rules: import_bcs.bcs.vector(import_bcs.bcs.string())
});
