// @mysten/kiosk@0.14.4/dist/cjs/constants.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var constants_exports = {};
__export(constants_exports, {
  FLOOR_PRICE_RULE_ADDRESS: () => FLOOR_PRICE_RULE_ADDRESS,
  KIOSK_LOCK_RULE_ADDRESS: () => KIOSK_LOCK_RULE_ADDRESS,
  PERSONAL_KIOSK_RULE_ADDRESS: () => PERSONAL_KIOSK_RULE_ADDRESS,
  ROYALTY_RULE_ADDRESS: () => ROYALTY_RULE_ADDRESS,
  getBaseRules: () => getBaseRules,
  mainnetRules: () => mainnetRules,
  rules: () => rules,
  testnetRules: () => testnetRules
});
module.exports = __toCommonJS(constants_exports);
var import_resolve = require("./tx/rules/resolve.js");
var import_types = require("./types/index.js");
const ROYALTY_RULE_ADDRESS = {
  [import_types.Network.TESTNET]: "bd8fc1947cf119350184107a3087e2dc27efefa0dd82e25a1f699069fe81a585",
  [import_types.Network.MAINNET]: "0x434b5bd8f6a7b05fede0ff46c6e511d71ea326ed38056e3bcd681d2d7c2a7879",
  [import_types.Network.CUSTOM]: ""
};
const KIOSK_LOCK_RULE_ADDRESS = {
  [import_types.Network.TESTNET]: "bd8fc1947cf119350184107a3087e2dc27efefa0dd82e25a1f699069fe81a585",
  [import_types.Network.MAINNET]: "0x434b5bd8f6a7b05fede0ff46c6e511d71ea326ed38056e3bcd681d2d7c2a7879",
  [import_types.Network.CUSTOM]: ""
};
const FLOOR_PRICE_RULE_ADDRESS = {
  [import_types.Network.TESTNET]: "0x06f6bdd3f2e2e759d8a4b9c252f379f7a05e72dfe4c0b9311cdac27b8eb791b1",
  [import_types.Network.MAINNET]: "0x34cc6762780f4f6f153c924c0680cfe2a1fb4601e7d33cc28a92297b62de1e0e",
  [import_types.Network.CUSTOM]: ""
};
const PERSONAL_KIOSK_RULE_ADDRESS = {
  [import_types.Network.TESTNET]: "0x06f6bdd3f2e2e759d8a4b9c252f379f7a05e72dfe4c0b9311cdac27b8eb791b1",
  [import_types.Network.MAINNET]: "0x0cb4bcc0560340eb1a1b929cabe56b33fc6449820ec8c1980d69bb98b649b802",
  [import_types.Network.CUSTOM]: ""
};
function getBaseRules({
  royaltyRulePackageId,
  kioskLockRulePackageId,
  personalKioskRulePackageId,
  floorPriceRulePackageId
}) {
  const rules2 = [];
  if (royaltyRulePackageId) {
    rules2.push({
      rule: `${royaltyRulePackageId}::royalty_rule::Rule`,
      packageId: royaltyRulePackageId,
      resolveRuleFunction: import_resolve.resolveRoyaltyRule
    });
  }
  if (kioskLockRulePackageId) {
    rules2.push({
      rule: `${kioskLockRulePackageId}::kiosk_lock_rule::Rule`,
      packageId: kioskLockRulePackageId,
      resolveRuleFunction: import_resolve.resolveKioskLockRule,
      hasLockingRule: true
    });
  }
  if (personalKioskRulePackageId) {
    rules2.push({
      rule: `${personalKioskRulePackageId}::personal_kiosk_rule::Rule`,
      packageId: personalKioskRulePackageId,
      resolveRuleFunction: import_resolve.resolvePersonalKioskRule
    });
  }
  if (floorPriceRulePackageId) {
    rules2.push({
      rule: `${floorPriceRulePackageId}::floor_price_rule::Rule`,
      packageId: floorPriceRulePackageId,
      resolveRuleFunction: import_resolve.resolveFloorPriceRule
    });
  }
  return rules2;
}
const testnetRules = getBaseRules({
  royaltyRulePackageId: ROYALTY_RULE_ADDRESS[import_types.Network.TESTNET],
  kioskLockRulePackageId: KIOSK_LOCK_RULE_ADDRESS[import_types.Network.TESTNET],
  personalKioskRulePackageId: PERSONAL_KIOSK_RULE_ADDRESS[import_types.Network.TESTNET],
  floorPriceRulePackageId: FLOOR_PRICE_RULE_ADDRESS[import_types.Network.TESTNET]
});
const mainnetRules = getBaseRules({
  royaltyRulePackageId: ROYALTY_RULE_ADDRESS[import_types.Network.MAINNET],
  kioskLockRulePackageId: KIOSK_LOCK_RULE_ADDRESS[import_types.Network.MAINNET]
});
const rules = [...testnetRules, ...mainnetRules];
