// @mysten/messaging@0.1.0/dist/cjs/encryption/constants.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var constants_exports = {};
__export(constants_exports, {
  ALLOWLISTED_SEAL_KEY_SERVERS: () => ALLOWLISTED_SEAL_KEY_SERVERS,
  ENCRYPTION_PRIMITIVES_CONFIG: () => ENCRYPTION_PRIMITIVES_CONFIG
});
module.exports = __toCommonJS(constants_exports);
const ENCRYPTION_PRIMITIVES_CONFIG = {
  keySize: 256,
  nonceSize: 12,
  dekAlgorithm: "AES-GCM"
};
const ALLOWLISTED_SEAL_KEY_SERVERS = {
  mainnet: [],
  testnet: [
    "0x73d05d62c18d9374e3ea529e8e0ed6161da1a141a94d3f76ae3fe4e99356db75",
    "0xf5d14a81a982144ae441cd7d64b09027f116a468bd36e7eca494f750591623c8"
  ]
};
